### PR TITLE
Add validation test, including a failing example

### PR DESCRIPTION
**Pull request by fumieval**
**Summary**

This PR adds a new test, `validation`, that tries to validate concrete examples of `SAMLResponse`s. It contains a couple of samples that wai-saml2 can't handle at the moment.

Most code is extracted from #45

**Checklist**

- [x] All definitions are documented with Haddock-style comments.
- [x] All exported definitions have `@since` annotations.
- [x] Code is formatted in line with the existing code.
- [x] The changelog has been updated.

### CODE REVIEW

File: package.yaml
Comment (mbg) [inline package.yaml]: Why the change?

File: package.yaml
Comment (mbg) [inline package.yaml]: I am curious about your decision to have this in an entirely separate test suite. What are your thoughts on having one test suite which combines all the tests with `tasty`?

File: tests/validation.hs
Comment (mbg) [inline tests/validation.hs]: Just noting for the record that these tests are expected to succeed once #45 is merged. No need to act on this comment.

File: CHANGELOG.md
Comment (mbg) [inline CHANGELOG.md]: This change probably doesn't need to be in the changelog, but I don't feel strongly either way. So happy for you to leave it in if you want.

**Review by mbg — COMMENTED** [review]
Thanks for moving this out! That does make it easier to review, although ironically I did already have pending review comments for this part of #45 🙈 so I have copied my relevant comments from there over to the review here. Sorry that other review has been taking much longer than I anticipated.

File: package.yaml
Comment (fumieval) [reply package.yaml]: Because GHC complains that the module name should be `Parser`, while I want this to be `Main`. That's probably because the tests directory now contains more than one file.

File: package.yaml
Comment (fumieval) [reply package.yaml]: That's a good question, I now think having an unified executable is better so I'll combine these tomorrow.

**Comment by fumieval** [comment]
I wonder why the tests fail on stack-lts-19 or older...

File: package.yaml
Comment (mbg) [inline package.yaml]: Minor: if we have one test suite we can follow the typical(?) naming convention for test suites:
```suggestion
  wai-saml2-test:
```

File: tests/Validation.hs
Comment (mbg) [inline tests/Validation.hs:52]: Before your most recent force push, this test was also guarded by `expectFail`:

```suggestion
    , expectFail $ testCase "Okta with AttributeStatement"
        $ run "okta.crt" "2023-06-16T06:43:00.000Z" "okta-attributes.xml"
```

I assume that this is why the CI fails for LTS 19 and older. The real question is why it _succeeds_ for LTS 20 when that should fail as well!

**Review by mbg — COMMENTED** [review]
It seems like you omitted an `expectFail` in your most recent commit, which I think leads to the test failures now. The real question is why the tests for LTS 20 don't fail 😕

File: tests/Validation.hs
Comment (fumieval) [reply tests/Validation.hs:52]: This is intentional because #51 is now merged and I rebased the branch on master (otherwise it would fail after merging),

Comment (fumieval) [reply tests/Validation.hs:52]: I commented out this test case for now

Comment (mbg) [reply tests/Validation.hs:52]: Ah I see, I hadn't realised that one of these test cases was for https://github.com/mbg/wai-saml2/pull/51. (See also my assumption that both tests were for https://github.com/mbg/wai-saml2/pull/45 in https://github.com/mbg/wai-saml2/pull/52#discussion_r1240718082).

It would be good for us to better understand why the test fails unexpectedly with the older Stack resolvers to at least establish whether this is an issue in one of the libraries we are using or some problem in our code. 

I am going to be on holiday from tomorrow for the next two weeks, so I don't have time right now to look more deeply into this, but I can do so when I am back if you want.

**Comment by fumieval** [comment]
I figured out the cause of the test failures, and fixed it in https://github.com/mbg/c14n/pull/3

File: tests/Validation.hs
Comment (fumieval) [reply tests/Validation.hs:52]: @mbg I updated `stack.yaml`s to use the latest c14n. That did the trick

File: stack-lts-16.1.yaml.lock
Comment (mbg) [inline stack-lts-16.1.yaml.lock:1]: AFAIK the `.lock` files are intended to be checked in to version control.

File: stack-lts-16.1.yaml
Comment (mbg) [inline stack-lts-16.1.yaml]: Could you make sure that the sha256 hash is included everywhere?

**Review by mbg — COMMENTED** [review]
Thanks again for figuring out and fixing the issue over in the `c14n` library that fixed the tests here. This mostly looks good now, just two minor niggles that would be good to sort out before merging this.

File: stack-lts-16.1.yaml
Comment (fumieval) [reply stack-lts-16.1.yaml]: Sure

**Comment by fumieval** [comment]
@mbg Friendly reminder that this is ready for review

**Review by mbg — APPROVED** [review]
Sorry about that! I must have missed that this was ready for another review; you can always hit the icon next to my username in the list of reviewers as well to re-request a review if you need to.

I think this looks good to merge as is now. Thanks for your work on finding and fixing the bug in `c14n`, and teasing these tests out into their own PR.

**Pull request merged**
